src/platform/plugins/shared/dashboard/public/dashboard_api/project_routing_manager.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import type { ProjectRouting } from '@kbn/es-query';
+import type { PublishingSubject, StateComparators } from '@kbn/presentation-publishing';
+import { diffComparators } from '@kbn/presentation-publishing';
+import type { Subscription } from 'rxjs';
+import { BehaviorSubject, combineLatestWith, debounceTime, map } from 'rxjs';
+import { cpsService } from '../services/kibana_services';
+import type { DashboardState } from '../../common';
+
+export const COMPARE_DEBOUNCE = 100;
+
+export function initializeProjectRoutingManager(
+  initialState: DashboardState,
+  projectRoutingRestore$: PublishingSubject<boolean>
+) {
+  if (!cpsService?.cpsManager) {
+    return;
+  }
+
+  const projectRouting$ = new BehaviorSubject<ProjectRouting>(initialState.project_routing);
+
+  function setProjectRouting(projectRouting: ProjectRouting) {
+    if (projectRouting !== projectRouting$.value) {
+      projectRouting$.next(projectRouting);
+    }
+  }
+
+  // pass the initial state to CPS manager from dashboard state or just reset to default on dashboard init
+  cpsService.cpsManager.setProjectRouting(
```

Comment:
Should this logic be moved into `setProjectRouting` so that setting `setProjectRouting` also updates `cpsService.cpsManager`? I see that `reset` also has to call `cpsService.cpsManager.setProjectRouting`. This looks like an easy way for these 2 values to get out of sync.